README.rst
==================
iw.rejectanonymous
==================


What is iw.rejectanonymous ?
============================

This package is made to reject unconditionnally anonymous users from a Plone
site, without any change in your security policy matrix or workflows. They
should get redirected by plone to login form. The basic use case is an extranet,
where all visitors must be authenticated.

Works with
==========

* Plone 3
* Plone 4
* Plone 5


Installation
============

Add ``iw.rejectanonymous`` to the ``eggs`` option of your
``plone.recipe.zope2instance`` part ::

  ...
  [instance]
  recipe = plone.recipe.zope2instance
  ...
  eggs =
      ...
      iw.rejectanonymous
      ...
  ...
  # The ZCML slug is no more required with Plone 3.3 and up
  zcml =
      ...
      iw.rejectanonymous
      ...

and re-run buildout.

To make a site private, you need to install `iw.rejectanonymous` and check
"Private site" flag on controlpanel.


Customization
=============

``iw.rejectanonymous`` enables the publication of some resources to the
anonymous user, more specifically to enable all media and resources required
from the standard loging page and the password reset page.

Adding valid ids
----------------

If your customized logging page requires some specific images or your site
policy component provides a signup page which name is not ``login_form`` you may
add additional ids (url last part) that are available to anonymous users.

::

  from iw.rejectanonymous import addValidIds
  ...
  addValidIds('some_image.png', 'my_login_form')


Adding valid subparts
---------------------

If you want to let anonymous users browse the pages of some folders, you need to
add valid subparts.

::

  from iw.rejectanonymous import addValidSubparts
  ...
  addValidSubparts('disclaimer', 'public_section')

Adding valid subparts prefixes
------------------------------

If you want to let anonymous users browse the pages of some folders with
specific prefixes, you need to add valid subpart prefixes.

::

  from iw.rejectanonymous import addValidSubpartPrefixes
  ...
  addValidSubpartPrefixes('public_')

Hiding viewlets
---------------

You may hide viewlets from the views of the site (login form, password reset
form). You need for this to add such lines in your site policy ZCML.

::

  <browser:viewlet
    name="original.viewlet.name"
    for="iw.rejectanonymous.IPrivateSite"
    manager="original.viewlet.manager.Interface"
    class="original.viewlet.Class"
    permission="cmf.SetOwnProperties"
  />

``name``
  Keep the original viewlet name.

``for``
  ``iw.rejectanonymous.IPrivateSite`` the marker interface set to private sites

``manager``
  Keep the original manager

``class``
  Keep the original viewlet class

``permission``
  Choose a permission that is not granted to an anonymous user but to anyone
  else. ``cmf.SetOwnProperties`` is a good choice if your site has the standard
  security policy.

See how we hide the ``plone.personal_bar`` and the ``plone.searchbox`` in the
``configure.zcml`` of this component.


Disable rejectanonymous
=======================

If you need to temporary disable rejectanonymous on your site and you don't want to
disable it globally (for example for debugging issues or some scheduled scripts),
you can set an environment variable `DISABLE_REJECTANONYMOUS`.

For example you can add it to your debug instance like this:

::

  environment-vars +=
      DISABLE_REJECTANONYMOUS true

Or pass it before running a script:

::

  > DISABLE_REJECTANONYMOUS true bin/instance run script foo.py


Links
=====

Cheeseshop
  http://pypi.python.org/pypi/iw.rejectanonymous

Git repository
  https://github.com/collective/iw.rejectanonymous

Issue tracker
  https://github.com/collective/iw.rejectanonymous/issues

Old SVN repository (up to 1.0.2)
  https://svn.plone.org/svn/collective/iw.rejectanonymous

Contributors
============

* Bertrand Mathieu
* Thomas Desvenain
* Gilles Lenfant
* Elisabeth Leddy


CHANGES.rst
Changes log
===========

2.0.0.dev1 (unreleased)
-----------------------

- Disable rejectanonymous also with an environment variable.
  [cekk]


2.0.0.dev0 (2017-01-01)
-----------------------

- Huge refactor and changes made to get the package work with Plone 5.1.
  [arsenico13]
- Compliancy with plone.rest: allow anonymous OPTIONS requests
  [ebrehault]


1.2.3 (2014-04-30)
------------------

- Modified coding style of getPortalLogoId function.
  [vincentfretin]

- Make iw.rejectanonymous work in tests
  [jaroel]


1.2.2 (2012-08-21)
------------------

- 4.2.1 Compatibility.
  [thomasdesvenain]


1.2.1 (2012-08-16)
------------------

- Include cmf permissions zcml.
  [thomasdesvenain]


1.2 (2012/5/16)
---------------

- ZCML duplicate viewlet setting
  [eleddy]

- Typos in README.rst
  [glenfant]

- Version in setup.py was a float
  [glenfant]


1.1 (2012/5/12)
---------------

- Added doc for customization
  [glenfant]

- Re enabled tests
  [glenfant]

- Disabled some viewlets to anonymous and added doc for customization.
  See https://github.com/collective/iw.rejectanonymous/issues/1
  [glenfant]

- Added valid subpart prefixes such iw.rejectanonymous does not conflict with
  plone.app.theming (Diazo)
  [glenfant]

- Enable portal_kss subparts.
  [thomasdesvenain]

- In plone 4, use of base_properties is sketchy and likely incompat.
  Catch error if needed and use better default for logo.
  [eleddy]


1.0.2 (2010-12-27)
------------------

- z3c.autoinclude awareness added so the ZCML slug does not need to be
  explicitely added in buildout ``*.cfg``.
  [glenfant]

- Add customization utilities and doc (add new enabled ids and subpaths)
  [glenfant]

- Enable favicon.
  [thomasdesvenain]


1.0.1 - 2010-10-08
------------------

- Enable password reset system.
  [thomasdesvenain]


1.0.0 - 2008-02-11
------------------

- Initial release
  [bmathieu]


iw/rejectanonymous/tests/base.py
# -*- coding: utf-8 -*-
# Copyright (C)2007 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
""" defines a test class and its Plone Site layer for plone tests
"""
from Products.PloneTestCase import PloneTestCase as ptc
from Products.PloneTestCase.layer import onsetup

from Products.Five import zcml
from Products.Five import fiveconfigure


@onsetup
def setup_reject_anonymous_site():
    """
    """
    import iw.rejectanonymous
    fiveconfigure.debug_mode = True
    zcml.load_config('configure.zcml', iw.rejectanonymous)
    fiveconfigure.debug_mode = False


setup_reject_anonymous_site()
ptc.setupPloneSite()


class TestCase(ptc.FunctionalTestCase):
    """test case used in tests"""


iw/rejectanonymous/tests/test_rejectanonymousdocs.py
# -*- coding: utf-8 -*-
# Copyright (C)2007 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""
Generic Test case for iw.rejectanonymous doctest
"""
__docformat__ = 'restructuredtext'

import unittest
import doctest
import sys
import os

from Testing.ZopeTestCase import FunctionalDocFileSuite
from Products.Five.testbrowser import Browser

from base import TestCase

import iw.rejectanonymous

current_dir = os.path.dirname(__file__)


def testSetUp(self):
    # self is a base.TestCase
    iw.rejectanonymous.valid_subparts = frozenset((
        'portal_css', 'portal_javascripts', 'passwordreset'
        ))


def doc_suite(test_dir, setUp=None, tearDown=None, globs=None):
    """Returns a test suite, based on doctests found in /doctest."""
    suite = []
    if globs is None:
        globs = globals()

    # preparing a few elements
    globs['test_dir'] = current_dir
    browser = Browser()
    browser.handleErrors = False
    globs['browser'] = browser

    flags = (doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE |
             doctest.REPORT_ONLY_FIRST_FAILURE)

    package_dir = os.path.split(test_dir)[0]
    if package_dir not in sys.path:
        sys.path.append(package_dir)

    docs = []
    for dir_ in ('doctests',):
        doctest_dir = os.path.join(package_dir, dir_)

        # filtering files on extension
        docs.extend([os.path.join(doctest_dir, doc)
                     for doc in os.listdir(doctest_dir)
                     if doc.endswith('.txt')])

    for test in docs:
        suite.append(FunctionalDocFileSuite(test, optionflags=flags,
                                            globs=globs, setUp=setUp,
                                            tearDown=tearDown,
                                            test_class=TestCase,
                                            module_relative=False))

    return unittest.TestSuite(suite)


def test_suite():
    """returns the test suite"""
    return doc_suite(current_dir, setUp=testSetUp)


if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


iw/rejectanonymous/setuphandlers.py
# -*- coding: utf-8 -*-
from iw.rejectanonymous import IPrivateSite
from iw.rejectanonymous import logger
from plone import api
from Products.CMFPlone.interfaces import INonInstallable
from Products.Five.utilities.marker import erase
from zope.interface import implementer


@implementer(INonInstallable)
class HiddenProfiles(object):

    def getNonInstallableProfiles(self):
        """Hide uninstall profile from site-creation and quickinstaller"""
        return [
            'iw.rejectanonymous:uninstall',
        ]


def post_install(context):
    """Post install script"""
    # Do something at the end of the installation of this package.


def uninstall(context):
    """Uninstall script"""
    site = api.portal.get()
    erase(site, IPrivateSite)
    logger.info("iw.rejectanonymous:uninstall - The site has been set public.")
    # Do something at the end of the uninstallation of this package.


iw/rejectanonymous/browser/plonecontrolpanel.py
# -*- coding: utf-8 -*-
# Copyright (C) 2008 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""Add a new "Private Site" control panel to manage the new option
"""
from iw.rejectanonymous import _
from iw.rejectanonymous import IPrivateSite
from iw.rejectanonymous import logger
from plone import api
from plone.app.registry.browser.controlpanel import ControlPanelFormWrapper
from plone.app.registry.browser.controlpanel import RegistryEditForm
from plone.supermodel import model
from Products.Five.utilities.marker import mark, erase
from z3c.form import button
from zope import schema


class IPrivateSiteSchema(model.Schema):
    private_site = schema.Bool(
        title=_(u'Private site'),
        description=_(u"Users must login to view the site. Anonymous users are presented the login form."),  # noqa
        default=False,
        required=False,
    )


class IPrivateSiteSettings(IPrivateSiteSchema):
    """
    Marker interface for settings
    """


class PrivateSiteSettingsEditForm(RegistryEditForm):
    schema = IPrivateSiteSettings
    label = _(u"Reject Anonymous settings")

    @button.buttonAndHandler(_(u"Save"), name='save')
    def handleSave(self, action):
        private_selected = not api.portal.get_registry_record('iw.rejectanonymous.browser.plonecontrolpanel.IPrivateSiteSettings.private_site')  # noqa
        site = api.portal.get()
        if private_selected:
            mark(site, IPrivateSite)
            logger.info("The site has been set private.")
        else:
            erase(site, IPrivateSite)
            logger.info("The site has been set public.")

        super(PrivateSiteSettingsEditForm, self).handleSave(self, action)

    @button.buttonAndHandler(_(u"Cancel"), name='cancel')
    def handleCancel(self, action):
        super(PrivateSiteSettingsEditForm, self).handleCancel(self, action)


class PrivateSiteSettingsView(ControlPanelFormWrapper):
    form = PrivateSiteSettingsEditForm


iw/rejectanonymous/__init__.py
# -*- coding: utf-8 -*-
# Copyright (C)2007 Ingeniweb

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING. If not, write to the
# Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
"""rejectanonymous initialization"""
from zope.interface import Interface
from AccessControl import getSecurityManager
from Acquisition import aq_get
from zExceptions import Unauthorized
from zope.i18nmessageid import MessageFactory

import logging
import os


logger = logging.getLogger("iw.rejectanonymous")
_ = MessageFactory('iw.rejectanonymous')


class IPrivateSite(Interface):
    """Marker for sites requiring login"""


valid_ids = frozenset((
    'login_form', 'require_login', 'login.js', 'spinner.gif',
    'mail_password_form', 'mail_password', 'contact-info', 'pwreset_form',
    'pwreset_finish', 'favicon.ico', 'logo.jpg', 'logo.png'
    ))

valid_subparts = frozenset((
    'portal_css', 'portal_javascripts', 'passwordreset', 'portal_kss'
    ))

valid_subpart_prefixes = frozenset(('++resource++', '++theme++'))


# Customization functions
def addValidIds(*new_ids):
    """A customized Plone site may need to publish other ids as resources
    of the login process. The policy or third party component just need to
    use this function for this to happen.

    :param new_ids: one or more ids
    """
    global valid_ids
    valid_ids |= set(new_ids)
    return


def addValidSubparts(*new_subparts):
    """A customized Plone site may need to publish other subparts for resources
    of the login process. The policy or third party component just need to
    use this function for this to happen.

    :param new_subparts: one or more traversal ids
    """
    global valid_subparts
    valid_subparts |= set(new_subparts)
    return


def addValidSubpartPrefixes(*new_prefixes):
    """A customized Plone site may need to publish other subpart prefixes for
    resources of the login process. The policy or third party component just
    need to use this function for this to happen.

    :param new_subparts: one or more traversal ids
    """
    global valid_subpart_prefixes
    valid_subpart_prefixes |= set(new_prefixes)


# Utilities
def isAnonymousUser():
    u = getSecurityManager().getUser()
    return (u is None or u.getUserName() == 'Anonymous User')


def getPortalLogoId(portal):
    props = aq_get(portal, 'base_properties', None)
    return props is not None and props.getProperty('logoName', '') or ''


def rejectAnonymous(portal, request):
    if request['REQUEST_METHOD'] == 'OPTIONS':
        return
    if os.environ.get('DISABLE_REJECTANONYMOUS'):
        return
    if isAnonymousUser():
        url = request.physicalPathFromURL(request['URL'])
        if url[-1] == 'index_html':
            url.pop()
        item_id = url[-1]
        logo_id = getPortalLogoId(portal)

        if url and not (
            item_id in valid_ids
            or item_id == logo_id
            or [path for path in url
                if path in valid_subparts]
            or [path for path in url
                if [v for v in valid_subpart_prefixes if path.startswith(v)]]
                ):
            raise Unauthorized, "Anonymous rejected"


def insertRejectAnonymousHook(portal, event):
    """ """
    try:
        event.request.post_traverse(rejectAnonymous, (portal, event.request))
    except RuntimeError:
        # Make this work in a testrunner
        pass

import iw.rejectanonymous.browser.plonecontrolpanel  # noqa
